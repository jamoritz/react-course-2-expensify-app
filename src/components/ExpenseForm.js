/*global console */

import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// // const date = new Date();
// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            calendarFocused: false,
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            description: props.expense ? props.expense.description : '',
            errorMsg: '',
            isAmountError: false,
            isDescriptionError: false,
            note: props.expense ? props.expense.note : ''
        };
        // console.log('ExpenseForm constructor', { ...this.state });
    }

    onAmountChange = (event) => {
        const amount = event.target.value;

        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onDescriptionChange = (event) => {
        const description = event.target.value;
        this.setState(() => ({ description }));
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState(() => ({ note }));
    };

    onSubmit = (event) => {
        event.preventDefault();

        let errorMsg = '';
        let isAmountError = false;
        let isDescriptionError = false;
        if (!this.state.description) {
            isDescriptionError = true;
            errorMsg += ' a description';
            if (!this.state.amount) {
                isAmountError = true;
                errorMsg += ' and an amount.';
            }
        } else if (!this.state.amount) {
            isAmountError = true;
            errorMsg += ' an amount.';
        }

        if (errorMsg) {
            this.setState(() => ({
                isAmountError: isAmountError,
                isDescriptionError: isDescriptionError,
                errorMsg: 'Please provide' + errorMsg
            }));
        } else {
            this.setState(() => ({
                isAmountError: false,
                isDescriptionError: false,
                errorMsg: ''
            }));

            this.props.onSubmit({
                amount: Math.floor(parseFloat(this.state.amount) * 100),
                createdAt: this.state.createdAt.valueOf(),
                description: this.state.description,
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.errorMsg && <p className="error-msg">{this.state.errorMsg}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        className={
                            this.state.isDescriptionError ? 'text-input error' : 'text-input'
                        }
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        className={
                            'text-input' + this.state.isAmountError
                                ? 'text-input error'
                                : 'text-input'
                        }
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        className="textarea"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button className="button">{this.props.action} Expense</button>
                </form>
            </div>
        );
    }
}
