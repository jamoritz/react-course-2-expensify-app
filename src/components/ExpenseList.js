/*global console */

import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// // forEach implementation -- does not generate ExpenseListItem output
// const ExpenseList = (props) => (
//     <div>
//         <h1>Expense List</h1>
//         {/*{console.log(props.expenses.length)}*/}
//         {props.expenses.forEach((expense) => {
//             // console.log(`forEach: ${expense.description}`);
//             return <ExpenseListItem key={expense.id} {...expense} />;
//         })}
//     </div>
// );

// map implementation -- does generate ExpenseListItem output
export const ExpenseList = (props) => (
    <div>
        {props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense} />;
            })
        )}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);
