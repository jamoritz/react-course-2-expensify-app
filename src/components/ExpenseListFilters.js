/*global console */

import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import {
    setEndDate,
    setStartDate,
    setTextFilter,
    sortByAmount,
    sortByDate
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setEndDate(endDate);
        this.props.setStartDate(startDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = (event) => {
        this.props.setTextFilter(event.target.value);
    };

    onSortChange = (event) => {
        event.target.value === 'amount' ? this.props.sortByAmount() : this.props.sortByDate();
    };

    render(props) {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            placeholder="Search expenses"
                            type="text"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
