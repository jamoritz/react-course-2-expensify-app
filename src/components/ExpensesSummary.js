/*global console */

import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount > 1 ? 'expenses' : 'expense';
    if (expenseCount > 0) {
        return (
            <div>
                <h1>
                    Viewing {expenseCount} {expenseWord} totalling {expensesTotal}
                </h1>
            </div>
        );
    }
    return (
        <div>
            <h1>Viewing no expenses</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: numeral(selectExpensesTotal(visibleExpenses) / 100).format('$0,0.00')
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
