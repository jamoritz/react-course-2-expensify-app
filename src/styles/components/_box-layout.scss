.box-layout {
    align-items: center;
    background: url('/images/bg.jpg');
    background-size: cover;
    display: flex;
    height: 100vh;
    justify-content: center;
    width: 100vw;

    &__box {
        background: rgba($dialog-bg-color, 0.85);
        border-radius: 0.3rem;
        padding: $large-size $medium-size;
        text-align: center;
        width: 25rem;
    }

    &__title {
        line-height: 1;
        margin: 0 0 $medium-size 0;
    }
}
