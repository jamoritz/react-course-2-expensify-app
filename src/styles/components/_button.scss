.button {
    background-color: $button-bg-color;
    border: none;
    color: $button-fg-color;
    display: inline-block;
    font-size: $font-size-large;
    font-weight: $font-weight-medium;
    line-height: 1;
    padding: $small-size;
    text-decoration: none;
    text-transform: capitalize;

    &--link {
        background-color: transparent;
    }
    &--secondary {
        background-color: $medium-grey;
    }
}
