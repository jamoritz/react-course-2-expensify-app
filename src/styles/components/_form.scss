.form {
    display: flex;
    flex-direction: column;
    > * {
        margin-bottom: $medium-size;
    }

    &__error {
        color: $error-fg-color;
        margin: 0 0 $medium-size 0;
        font-style: italic;
    }
}