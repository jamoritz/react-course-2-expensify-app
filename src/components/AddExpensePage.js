/*global console */

import React from 'react';
import { connect } from 'react-redux';

import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    };

    render = () => (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm action="add" onSubmit={this.onSubmit} />
        </div>
    );
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
