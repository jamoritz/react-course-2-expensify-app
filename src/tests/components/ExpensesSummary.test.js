/*global console, expect */

import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';

// test('should ', () => {});

test('it should render the ExpensesSummary component with 3 expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={'$1362.47'} />);
    expect(wrapper).toMatchSnapshot();
});

test('it should render the ExpensesSummary component with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={'$32.95'} />);
    expect(wrapper).toMatchSnapshot();
});

test('it should render the ExpensesSummary component with no expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0} />);
    expect(wrapper).toMatchSnapshot();
});
