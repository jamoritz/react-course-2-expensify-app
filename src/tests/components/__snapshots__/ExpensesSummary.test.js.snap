// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render the ExpensesSummary component with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling
       
      <span>
        $32.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`it should render the ExpensesSummary component with 3 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling
       
      <span>
        $1362.47
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`it should render the ExpensesSummary component with no expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing no expenses
    </h1>
  </div>
</div>
`;
